### PR TITLE
Architecture: Refactor Non fungible Vault + Many Refactors

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
* Add Index API
* Refactor NonFungible vaults to use Index API
* Add 3 variants of `SubstateKey`: `Tuple`, `Map`, `Sorted`
* Refactored Substate lock api to be more specific given the datastructure
* Removed receiver from method invocation and only allow tuple substates to be accessed by method actor

## Details
An Index API has been added to support iterable substates specifically for use with NonFungibles. This lets us get rid of any memory problems associated with using a `BTreeMap` to store non fungibles.

A separation of abstraction between a `DBKey` and a `SubstateKey` has also been added. This is in order to support some constraints which a database may require which we don't want to necessarily be constrained with at the substate layer. An example of this is JMTs requirement that keys are spread evenly.

### For Internal Integrators
The substate layout of NonFungible vaults has been updated to use an index.

### CODE REVIEW

File: radix-engine/src/track/track.rs [hunk 1]
```diff
@@ -114,22 +110,106 @@ impl TrackedSubstateKey {
...
+            | TrackedSubstateKey::ReadExistAndWrite(_, Write::Delete)
+            | TrackedSubstateKey::ReadOnly(ReadOnly::NonExistent)
+            | TrackedSubstateKey::Garbage => None,
+        }
+    }
+
+    pub fn set(&mut self, value: IndexedScryptoValue) {
+        match self {
+            TrackedSubstateKey::Garbage => {
+                *self = TrackedSubstateKey::WriteOnly(Write::Update(RuntimeSubstate::new(value)));
```
Comment (iamyulong) [inline radix-engine/src/track/track.rs:139]: Can it be `ReadAndWrite` here?

File: radix-engine/src/system/system.rs [hunk 2]
```diff
@@ -665,24 +582,200 @@ where
...
+            .kernel_remove_substate(node_id, module_id, &substate_key)?
+            .map(|v| v.into());
+
+        Ok(rtn)
+    }
+
+    fn scan_index(&mut self, node_id: &NodeId, count: u32) -> Result<Vec<Vec<u8>>, RuntimeError> {
+        let type_info = TypeInfoBlueprint::get_type(&node_id, self.api)?;
+        match type_info {
+            TypeInfoSubstate::IterableStore => {}
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:733]: Should we rename all `IterableStore` to `Index`, or the other way, for consistency?

File: radix-engine/src/system/system.rs [hunk 3]
```diff
@@ -699,7 +792,7 @@ where
     }
 
     #[trace_resources]
-    fn insert_into_sorted_store(
+    fn insert_into_sorted_index(
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:795]: Same for `SortedStore`

**Review by iamyulong — APPROVED** [review]
LGTM with some minor comments.

File: radix-engine/src/system/system.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:733]: Yup good catch

File: radix-engine/src/track/track.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/track/track.rs:139]: Nope, this should never be `ReadAndWrite` as `Garbage` is only created from `WriteOnly`

File: simulator/src/ledger/dumper.rs [hunk 4]
```diff
@@ -114,19 +111,6 @@ pub fn dump_component<T: SubstateDatabase, O: std::io::Write>(
         );
     }
 
-    writeln!(output, "{}", "Non-fungible Resources".green().bold());
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply simulator/src/ledger/dumper.rs:117]: Added back in

**Pull request merged**